### ISSUE TITLE
Reloading state data

### ISSUE THREAD

**Issue by baj84** [AngularClass/NG6-starter#111]
Hey,

What would be the best way to reload the data for a component? We currently have many components on a page and am unsure on the best way to allow the user to go back and fourth on the browser and keeping the state data persistent.

Cheers

**Comment by fesor** [comment]
store application state somewere in services, get it using resolvers, pass to components via bindings, change state by asking services.

**Comment by samithaf** [comment]
I am using Angular cache factory within the service layer to solve this issue.

**Comment by fesor** [comment]
@samithaf could you share with more details?

**Comment by samithaf** [comment]
Sorry for the late response.

First you can register your cache factory as follows. 
```
.factory('app.core.common.mario.marioCache', $cacheFactory => {
    'ngInject'
    return $cacheFactory('app.core.common.mario.marioCache')
  })
```

and then you can write a service as follows which will consume the cache factory.

```
class MarioService {
  constructor ($http, $log, $rootScope, $cacheFactory) {
    this.$http = $http
    this.$log = $log
    this.$rootScope = $rootScope
    this.$cacheFactory = $cacheFactory

    this.model = {}
  }

getDetails(){
// pass the cache factory to http get. So Angular going to cache the data
 return this.$http.get('api/get/queen', { cache: this.$cacheFactory }).then((res) => {})
}

}
MarioService.$inject = ['$http', '$log', '$rootScope', 'app.core.common.mario.marioCache']

export default MarioService
```

**Comment by fesor** [comment]
@samithaf well... this doesn't solves any issues. This is just http response caching.

**Comment by samithaf** [comment]
@fesor by setting cache to http object when you request the same URL angular will not going to send another request. The original question is how to prevent data get loading from server again and again isn't it? So by adding cache problem can be solved.

**Comment by fesor** [comment]
@samithaf nope. Imagine that there is no backend. All state is stored in memory. And if you have many services, you could easily lose control and break state consistency. This is common problem for stateful components.

It's more about where to store state, in components or elsewhere. From where components should get state? Should it require state from services directly or there is a better way?

My answers was that components should be stateless and all required state should be provided from outside via bindings. Components shouldn't change application state directly, instead of it, they should ask service layer.

**Comment by slurmulon** [comment]
@baj84 if you are struggling with state a lot, I have written an Angular module that will only delegate state updates when necessary (via PubSub) to your contextually relevant entities, giving you more control over when entities are updated:

https://github.com/slurmulon/ng-current

I have used a similar approach in production - they idea is reacting to changes when necessary instead of relying so heavily on the synchronization of the digest cycle through rapid polling (this is something I have always felt Angular 1.X did wrong, but was right for the time).  As soon as we integrated this basic concept, issues such as state drift, missed updates and unwanted state refreshes became a thing of the past as their management became more transparent.

**Comment by fesor** [comment]
@slurmulon are you receiving state inside components?

**Comment by slurmulon** [comment]
@fesor Yes, a directive/controller is where you ultimately receive your states. Check out this demo code for an example:

https://github.com/slurmulon/ng-current/tree/master/demo/services (contextual services)
https://github.com/slurmulon/ng-current/blob/master/demo/directives.js (components incorporating contextual services)

For a working example, check out http://plnkr.co/edit/XlQ9ho?p=preview (particularly `script.js:L248-290`, helps to outline the general use case through data). The idea is that you commonly work with a hierarchy of related component entities (`User -> Site -> Quote`), each of which may be "selected" by a user at a certain time. For example, a user may have multiple `Sites` (perhaps a clearer term is `ConstructionSite` for this example, whatever), but you need to track their selected choice so that more detailed information may be displayed later on in a component.  If the user switches a `Site`, you might want any dependents (`Quote`) to be updated as well.

Note how when you click "Switch Users", the collection of `Sites` and `Quotes` represents those only belonging to the contextually relevant user. When you click "Random Site", only the list of quotes changes, and only if the newly chosen random site is different from the previous (intentional to show how this only delegates changes when necessary)

**Comment by fesor** [comment]
@slurmulon there are much simpler and cleaner solution of your problem than your implementation. All that we need to do - move out state manipulation outside of components and life became much easier:

http://plnkr.co/edit/fJP3akbbngumTOPygBPG?p=preview

p.s. Stop using scope, really. And please follow angular-styleguide.

**Comment by slurmulon** [comment]
I happily accept cleaner solutions and I appreciate your example. However I have worked with many great Angular/SPA engineers for nearly 6 years now who have struggled with handling the various implications of this problem, so this is simply a result of those experiences.

I view that there are a couple undesired complications with your example.

1. Your code does not allow you to easily abstract out business/model logic for `Site` or `Quote` with discrete services as Angular encourages. It assumes that you should explicitly access and manage every sub-entity of `User`, which doesn't encourage clean abstraction or proper encapsulation as far as I see it.

2. You must manually delegate the updates to other dependent entities **wherever they are used**, logic which you have chosen to reduce to:

   ```javascript
  onRandUserPressed() {
    this.User.all()
      .then(users => randomItem(users))
      .then(user => {
        this.currentUser = user;
        this.sites = user.sites;
      })
      .then(() => this.onRandSitePressed())
  }
 ```

 The entire point of this library is to prevent redundant code such as this because it's repetitive, prone to error and exposes complexity where you least want it. This logic must be re-written (or alternatively abstracted as I've done) in **every component** that involves either selecting or updating a `User`, `Site`, or `Quote` entity, which in the UIs I have worked on often involve many inter-dependent components. In other words, with this design, you must always remember that any time you switch a `User`, you must also switch a `Site`. In this particular snippet, I also find it strange that you must chain `.then(() => this.onRandSitePressed())` as this functionality should be orthogonal to the functionality of the "Switch User" button - seems to be a mix of concerns.

 These issues will be exacerbated when you need to coordinate the source of your data between `cache` and the latest API representation. For instance, what if your API supports representing resource entities as either links, partial or full responses? The logic you're suggesting easily breaks down under this common scenario because it strongly depends on a canonical source of data in order for it to stay clean and maintainable. Contrary to your statement, I do not believe that this code is easily maintainable in a large modern application.

If you have encountered complex problems such as these and elegantly solved them in the past, I'd be curious to see how you did so. Thanks for the feedback!

**Comment by fesor** [comment]
> Your code does not allow you to easily abstract out business/model logic for Site or Quote with discrete services as Angular encourages.

I prefer to place such logic in my domain entities. I trying to make services stateless.

> The entire point of this library is to prevent redundant code such as this because it's repetitive

In my projects I use ui-router's resolvers to get application state. In that case I would have 3 resolvers. Separate services which know how to get application state and nothing more. So... what's wrong with separation of concerns?

>  in every component that involves either selecting or updating a User, Site, or Quote entity

Components doesn't know anything about how to get or update entities. They are want state passed in and services to ask for update. Components should be stupid and stateless, and then we won't have problems with it.

> I also find it strange that you must chain .then(() => this.onRandSitePressed()) as this functionality should be orthogonal to the functionality of the "Switch User" button - seems to be a mix of concerns.

Well.. this could be solved in other way:

```
class MyComponent {
    // ...
    get user() { return this._user }
    set user(val) {
        this._user = val;
        this.onRandSitePressed();
    }
}
```

And since "sites" are references to users, they also should be updated after switching user. Or maybe this is some strange business logic.

> Contrary to your statement, I do not believe that this code is easily maintainable in a large modern application.

The main point is to use stupid stateless UI components. They have only one responsibility - representing state. The should not know where this state came from or how to update it. Only service layer know this. (my app component know how to get state just because we don't have router in this example)

I tried solutions like yours year ago or so, and it just doesn't work. It hard to test, it hard to work with. In my case, there was no problems so far.

I will try to update some opensouce app (docker distribution frontend for example) to my approach, maybe this would help.

**Comment by slurmulon** [comment]
> >Your code does not allow you to easily abstract out business/model logic for Site or Quote with discrete services as Angular encourages.

> I prefer to place such logic in my domain entities. I trying to make services stateless.

But by making your `Service`s "stateless", you have no easy way to share/delegate that state without propagating it through a chain of dependencies, scope bindings or verbose state providers. I agree that statelessness should be generally desired, it has many benefits. But ultimately there are many common scenarios where the changes in one component can affect any number of other component states. How would you easily orchestrate something like this without some sort of high-level supervisor (I guess I'm thinking of this under the Actor paradigm)? My solution was to look at each `Service` as a synchronized and canonical source of state for your entities, and this has worked quite well for me.  I'm definitely curious if you have avoided issues such as this entirely in the past, but from everything I've learned about state, concurrency, functional parallelism, etc., it seems you would experience some design complications without such singleton entities as the application grows.

---

> >The entire point of this library is to prevent redundant code such as this because it's repetitive
> In my projects I use ui-router's resolvers to get application state. In that case I would have 3 resolvers. Separate services which know how to get application state and nothing more. So... what's wrong with separation of concerns?

I suppose I just view it as pointless/irritating to involve entities such as this - everything has a sense of state regardless if the underlying mechanisms are stateless, even a web API, so it feels strange to abstract that in a decoupled manner.  Another problem with this is that anywhere a `Service` is used, I must be sure to `resolve` against that state's route with the latest and greatest, which seems cumbersome and repetitive to me (but admittedly free, you won't run into walls this way).  As I understand it, if you use your `Service` in 10 different components, which each have their own state, you would have to write `10` resolve blocks. I have worked on applications with nearly 100 directives/components, each of which may involve multiple `Service`s, so that could easily add up to hundreds of identical `resolve` statements (and their associated state provider statements) being made. Using `angular-ui-router` on large code base just seemed to complicate my state management scenario more than resolve it (the entire team felt this way), particularly because the applications I've worked on had a lot of complicated states which could not always be represented via URLs, forcing me to partially diverge from the framework and thus a canonical source of state (sad times). I also had problems with making certain data accessible in nested components as your state always stems from a high-level component that's mapped to a route. If you're using isolate scopes, which is common, the verbosity and duplication is incredibly frustrating.

---

>>I also find it strange that you must chain .then(() => this.onRandSitePressed()) as this functionality should be orthogonal to the functionality of the "Switch User" button - seems to be a mix of concerns.
> Well.. this could be solved in other way:

> ```
> class MyComponent {
>     // ...
>     get user() { return this._user }
>     set user(val) {
>         this._user = val;
>         this.onRandSitePressed();
>     }
> }
> ```
And since "sites" are references to users, they also should be updated after switching user. Or maybe this is some strange business logic.

I still feel that there is too much manual work involved here that is prone to mistake.  Why wouldn't you prefer something like:

```javascript
class RandomUser
  constructor(User) {
    'ngInject'
    
    User.use('current', then((user) => this.user = user)
  }

  next() {
    return User.all(function(users) {
      const current = Contexts.current(user, users[0]) // arbitrarily select first user if none is selected, can come from anywhere
      const others = users.filter((user) => user.id !== current.id) // users that aren't the current one

      if (others) {
        Contexts.select('user', others[0]) // select next user in the list, delegating update to all dependent Services and components
      }
    })
  }
```

In this way, you don't have to needlessly invoke `this.onRandSitePressed();`, it only ever gets called when it needs to by the user, and this guarantees that this component will always be working with the application's relevant user selection and never anything else.

---

>> Contrary to your statement, I do not believe that this code is easily maintainable in a large modern application.
> The main point is to use stupid stateless UI components. They have only one responsibility - representing state. The should not know where this state came from or how to update it. Only service layer know this. (my app component know how to get state just because we don't have router in this example)

I couldn't agree more, stupid stateless UI components should be preferred. `ng-current` doesn't break that paradigm that at all since the component utilizing it doesn't have to know/care where the state came from, it just needs to concern itself with the semantic of the state (i.e, what does this state mean to me?), which is an appropriate consideration in my eyes. All it needs to know is what entity it depends on and what the context of that dependency is, which is exactly what `ng-current` encourages (just inject `User` if you want it, then call `User.all()`, `User.byId()`, `User.current()`, whatever makes sense for the context of the component).  It's no different than pushing this logic up into the router and having a state provider determine this context instead - the latter just involves more user code. I believe your approach provides the same behavior, however it seems to involve more explicit state setup and repeated bindings in order for it to function, which is what I want to avoid. How have you addressed managing the state of components that don't fit into the URL scheme?

---

> I tried solutions like yours year ago or so, and it just doesn't work. It hard to test, it hard to work with. In my case, there was no problems so far.

> I will try to update some opensouce app (docker distribution frontend for example) to my approach, maybe this would help.

Interestingly I have experienced the opposite, testing became sane, incredibly predictable and easy to stub once we integrated these concepts. What tools/approaches did you use in the past that were similar to `ng-current`? I'm also curious how you cleanly coordinated cache and API representations with your approach.

Thanks for putting together more resources, really interested to take a look.
